lib/Dialect/FIRRTL/Transforms/GrandCentral.cpp

```diff
@@ -13,548 +13,1055 @@
 #include "PassDetails.h"
 #include "circt/Dialect/FIRRTL/FIRRTLAnnotations.h"
 #include "circt/Dialect/FIRRTL/FIRRTLVisitors.h"
+#include "circt/Dialect/FIRRTL/InstanceGraph.h"
 #include "circt/Dialect/FIRRTL/Passes.h"
 #include "circt/Dialect/HW/HWDialect.h"
 #include "circt/Dialect/SV/SVOps.h"
+#include "circt/Support/LLVM.h"
 #include "mlir/IR/ImplicitLocOpBuilder.h"
 #include "llvm/ADT/StringSwitch.h"
+#include "llvm/Support/Debug.h"
+#include <variant>
+
+#define DEBUG_TYPE "gct"
 
 using namespace circt;
 using namespace firrtl;
+using llvm::Optional;
+using llvm::SmallString;
+using llvm::StringSwitch;
 
 //===----------------------------------------------------------------------===//
 // Pass Implementation
 //===----------------------------------------------------------------------===//
 
 namespace {
-/// Mutable store of information about an Element in an interface.  This is
-/// derived from information stored in the "elements" field of an
-/// "AugmentedBundleType".  This is updated as more information is known about
-/// an Element.
-struct ElementInfo {
-  /// Encodes the "tpe" of an element.  This is called "Kind" to avoid
-  /// overloading the meaning of "Type" (which also conflicts with mlir::Type).
+
+/// Base class of all AugmentedTypes.  Roughly, this is a FIRRTL type that may
+/// have some more information about it.  It may also contain other
+/// AugmentedTypes.  Alternatively, this can be viewed as a mechanism to provide
+/// structure over a DictionaryAttr which is actually representing an
+/// AugmentedType.
+struct AugmentedType {
+
   enum Kind {
-    Error = -1,
+    Unknown = -1,
     Ground,
     Vector,
     Bundle,
     String,
     Boolean,
     Integer,
-    Double
+    Double,
+    Literal,
+    Deleted
   };
-  /// The "tpe" field indicating if this element of the interface is a ground
-  /// type, a vector type, or a bundle type.  Bundle types are nested
-  /// interfaces.
-  Kind tpe;
-  /// A string description that will show up as a comment in the output Verilog.
-  StringRef description;
-  /// The width of this interface.  This is only non-negative for ground or
-  /// vector types.
-  int32_t width = -1;
-  /// The depth of the interface.  This is one for ground types and greater
-  /// than one for vector types.
-  uint32_t depth = 0;
-  /// Indicate if this element was found in the circuit.
-  bool found = false;
-  /// Trakcs location information about what was used to build this element.
-  SmallVector<Location> locations = {};
-  /// True if this is a ground or vector type and it was not (statefully) found.
-  /// This indicates that an interface element, which is composed of ground and
-  /// vector types, found no matching, annotated components in the circuit.
-  bool isMissing() { return !found && (tpe == Ground || tpe == Vector); }
+
+private:
+  const Kind kind;
```

Comment:
Maybe my answer is below, but could you just make a new type hirarchy in tablegen for this?

Reply:
You're probably right. This occurred to me, but I already had this solution like 90% of the way there and just finished it this way. I'm pretty sure that everything I've written here is _exactly_ what tablegen is doing automatically.

Reply:
I cleaned this up to use new attributes in a tablegen hierarchy. There are probably further improvements in both the implementation in tablegen and switching to scattering the attributes defined in the tablegen hierarchy.